src/System.Text.Json/tests/Serialization/CustomConverterTests.NullValueType.cs

```diff
@@ -59,5 +61,166 @@ public static void ValueTypeConverterForNullWithArray()
+            public DateTimeOffset? NullableValue { get; set; }
+        }
+
+        [Fact]
+        public static void ValueConverterForNullableWithJsonConverterAttribute()
+        {
+            ClassWithNullableAndJsonConverterAttribute obj;
+
+            const string BaselineJson = @"{""NullableValue"":""1989/01/01 11:22:33""}";
+            obj = JsonSerializer.Deserialize<ClassWithNullableAndJsonConverterAttribute>(BaselineJson);
+            Assert.NotNull(obj.NullableValue);
+
+            const string Json = @"{""NullableValue"":""""}";
+            obj = JsonSerializer.Deserialize<ClassWithNullableAndJsonConverterAttribute>(Json);
+            Assert.Null(obj.NullableValue);
+
+            string json = JsonSerializer.Serialize(obj);
+            Assert.Contains(@"""NullableValue"":null", json);
+        }
+
+        private class ClassWithNullableAndWithoutJsonConverterAttribute
+        {
+            public DateTimeOffset? NullableValue { get; set; }
+            public List<DateTimeOffset?> NullableValues { get; set; }
+        }
+
+        [Fact]
+        public static void ValueConverterForNullableWithoutJsonConverterAttribute()
+        {
+            const string Json = @"{""NullableValue"":"""", ""NullableValues"":[""""]}";
+            ClassWithNullableAndWithoutJsonConverterAttribute obj;
+
+            // The json is not valid with the default converter.
+            Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<ClassWithNullableAndWithoutJsonConverterAttribute>(Json));
+
+            JsonSerializerOptions options = new JsonSerializerOptions();
+            options.Converters.Add(new JsonNullableDateTimeOffsetConverter());
+
+            obj = JsonSerializer.Deserialize<ClassWithNullableAndWithoutJsonConverterAttribute>(Json, options);
+            Assert.Null(obj.NullableValue);
+            Assert.Null(obj.NullableValues[0]);
+
+            string json = JsonSerializer.Serialize(obj);
+            Assert.Contains(@"""NullableValue"":null", json);
+            Assert.Contains(@"""NullableValues"":[null]", json);
+        }
+
+        [JsonConverter(typeof(ClassThatCanBeNullDependingOnContentConverter))]
+        private class ClassThatCanBeNullDependingOnContent
+        {
+            public int MyInt { get; set; }
+        }
+
+        /// <summary>
+        /// Allow a conversion of ClassThatCanBeNullDependingOnContent to null when its MyInt property is 0.
+        /// </summary>
+        private class ClassThatCanBeNullDependingOnContentConverter : JsonConverter<ClassThatCanBeNullDependingOnContent>
+        {
+            public override ClassThatCanBeNullDependingOnContent Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                if (reader.TokenType == JsonTokenType.Null)
+                {
+                    return null;
+                }
+
+                // Assume a single property.
+
```

Comment:
nit: assert token type is StartObject to be consistent with below assertions?

Reply:
The asserts in this converter are unnecessary since the semantics are verified by the calling test. I added the other assert and comment for readability.